crates/storage/trie/node/leaf.rs

```diff
@@ -0,0 +1,340 @@
+use crate::{
+    error::StoreError,
+    trie::{
+        db::TrieDB,
+        hashing::{NodeHash, NodeHashRef, NodeHasher, PathKind},
+        nibble::NibbleSlice,
+        node::BranchNode,
+        PathRLP, ValueRLP,
+    },
+};
+
+use super::{ExtensionNode, Node};
+/// Leaf Node of an an Ethereum Compatible Patricia Merkle Trie
+/// Contains the node's hash, value & path
+#[derive(Debug, Clone, Default)]
+pub struct LeafNode {
+    pub hash: NodeHash,
+    pub path: PathRLP,
+    pub value: ValueRLP,
+}
+
+impl LeafNode {
+    /// Creates a new branch and stores the given (path, value) pair
+    pub fn new(path: PathRLP, value: ValueRLP) -> Self {
+        Self {
+            hash: Default::default(),
+            path,
+            value,
+        }
+    }
+
+    /// Returns the stored value if the given path matches the stored path
+    pub fn get(&self, path: NibbleSlice) -> Result<Option<ValueRLP>, StoreError> {
+        if path.cmp_rest(&self.path) {
+            Ok(Some(self.value.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Stores the received value and returns the new root of the subtrie previously consisting of self
+    pub fn insert(
+        mut self,
+        db: &mut TrieDB,
+        path: NibbleSlice,
+        value: ValueRLP,
+    ) -> Result<Node, StoreError> {
+        /* Possible flow paths:
+            Leaf { SelfPath, SelfValue } -> Leaf { SelfPath, Value }
+            Leaf { SelfPath, SelfValue } -> Branch { [ Leaf { Path, Value }, Self, ... ], None, None}
+            Leaf { SelfPath, SelfValue } -> Extension { Branch { [Self,...] Path, Value } }
+            Leaf { SelfPath, SelfValue } -> Extension { Branch { [ Leaf { Path, Value } , ... ], SelfPath, SelfValue} }
+        */
+        self.hash.mark_as_dirty();
+        // If the path matches the stored path, update the value and return self
+        if path.cmp_rest(&self.path) {
+            self.value = value;
+            Ok(self.clone().into())
```

Comment:
We never mutate nodes stored in the DB so we don't overwrite historical data. The node returned by the method becomes the new root of the trie (or subtrie), which means that it will replace the current node in the cannonical trie.